src/coreclr/jit/gentree.cpp

```diff
@@ -14200,13 +14200,13 @@ GenTree* Compiler::gtFoldExprConst(GenTree* tree)
 
                             case TYP_FLOAT:
                             case TYP_DOUBLE:
-                                if (tree->IsUnsigned() && (lval1 < 0))
+                                if (tree->IsUnsigned())
                                 {
-                                    d1 = FloatingPointUtils::convertUInt64ToDouble((unsigned __int64)lval1);
+                                    d1 = FloatingPointUtils::convertUInt64ToDouble((uint64_t)lval1);
```

Comment:
I note that this code (and in VN) is equivalent to the previous version as we do not fold overflowing conversions.

Were we to start doing so, I note, I think this code becomes incorrect: the `GT_CAST` in our new model as I understand it is the "target instruction" which produces sentinels on overflow... sometimes. Sometimes it is equivalent to lowered code which produces ".NET" results, not "platform" results.

So, all in all, I don't know, more `GT_CAST` goodness for us I guess 😄. Leave the code as-is perhaps?

Reply:
Right. This is one of the things I still need to handle before moving it out of draft mode.

Reply:
Yep, this causes a break with optimizations on. For `double`, the tests I added pass with optimizations disabled, but the following fail (as no fallback is emitted) otherwise:
```
Unchecked conversion for Double -> System.Int32 failed; Input: ∞; Expected 2147483647; Actual -2147483648
Unchecked conversion for Double -> System.Int32 failed; Input: 2147483648; Expected 2147483647; Actual -2147483648
Unchecked conversion for Double -> System.Int64 failed; Input: NaN; Expected 0; Actual -9223372036854775808
Unchecked conversion for Double -> System.Int64 failed; Input: ∞; Expected 9223372036854775807; Actual -9223372036854775808
Unchecked conversion for Double -> System.Int64 failed; Input: 9.223372036854776E+18; Expected 9223372036854775807; Actual -9223372036854775808
Unchecked conversion for Double -> System.Int64 failed; Input: 9.223372036854776E+18; Expected 9223372036854775807; Actual -9223372036854775808
Unchecked conversion for Double -> System.UInt64 failed; Input: NaN; Expected 0; Actual 9223372036854775808
Unchecked conversion for Double -> System.UInt64 failed; Input: ∞; Expected 18446744073709551615; Actual 9223372036854775808
Unchecked conversion for Double -> System.UInt64 failed; Input: 1.8446744073709552E+19; Expected 18446744073709551615; Actual 9223372036854775808
Unchecked conversion for Double -> System.UInt64 failed; Input: 1.8446744073709552E+19; Expected 18446744073709551615; Actual 9223372036854775808
```

For `float`, with optimizations disabled there are still the following failures to work out:
```csharp
Unchecked conversion for Single -> System.UInt32 failed; Input: ∞; Expected 4294967295; Actual 0
Unchecked conversion for Single -> System.UInt32 failed; Input: 4.2949673E+09; Expected 4294967295; Actual 0
Unchecked conversion for Single -> System.UInt32 failed; Input: 4.294967E+09; Expected 4294967295; Actual 4294967040
Checked conversion for Single -> System.UInt32 failed; Input: 4.294967E+09; Expected 4294967295; Actual 4294967040
Unchecked conversion for Single -> System.UInt32 failed; Input: 4.2949673E+09; Expected 4294967295; Actual 0
```

Reply:
I expect the issue for `TYP_FLOAT -> TYP_UINT` is because it's currently using `TYP_LONG` and it needs to instead emit its own helper logic to account for `4294967295.0` to being exactly representable.